### ISSUE TITLE
Add gatsby-mdx-embed

### ISSUE THREAD

**Issue by ehowey** [ehowey/gatsby-theme-catalyst#352]
Wait until this is stable and then add in as a dependency.

**Comment by osiux** [comment]
What would you think on adding [gatsby-remark-embedder](https://github.com/MichaelDeBoey/gatsby-remark-embedder) instead? It's pretty stable and lots of providers are being added :) haven't tested it with MDX, but I think i should work.

**Comment by ehowey** [comment]
I have been watching that one too. They are very similar. Do you have experience with either that you can add?

The point that @pauliescanlon made on Twitter about the ability to shadow the mdx components in gatsby-mdx-embed is an important one. With gatsby-remark-embedder I am not sure how much flexibility theme authors would have to restyle or reconfigure the components if needed whereas mdx-embed is setup so that components could be shadowed. That is where my thinking is at right now anyways.

**Comment by PaulieScanlon** [comment]
Hey @ehowey! How's it going? 

`gatsby-remark-embedder` and `gatsby-mdx-embed` effectively do the same thing. 

They pretty much come down to two things.

1. Do you prefer to copy and paste a provider url in your `.md`
or 
2. Are you ok to manually extract and supply part of the provider url as a prop to a component.

... and as you've mentioned the ability to shadow the components could prove useful.

...lastly if you're  ok with everything being `display: block;` with no side by side layouts or responsive breakpoints markdown will do the job just fine.

My thinking behind creating `gatsby-mdx-embed` is that because it's under pinned by `.mdx` we open ourselves up to what is basically React development with some markdown thrown in for good measure. 

In the next iteration of the project i plan to bring the copy/paste url support to a more generic component

`<MdxEmbed url='' />`

Under the hood it would do all the clever url extraction and ensure the correct embed providers embed code is surfaced in the browser, thus making it easier to use in real life.

Also, with the use of something like `theme-ui/components` in `.mdx` you could use the `<Flex />` and `<Box /> components which act as a sort of bootstrap row / col setup, and in any of the Box(s) you could put a `gatsby-mdx-embed` component and give the author full control of the page layout in both desktop and mobile scenarios 

As mentioned above with `.md` you're pretty much always gonna be `display: block;` with content flowing one below the other as per it's `relative` position

As `.mdx` says itself ... it's for ambitious projects 

In my opinion `.mdx` offers all the great things we know and love about markdown and then allows us all the things we'd normally do in React/HTML/Css too.

**Comment by ehowey** [comment]
Great explanation Paul! That really helps cement my thinking on your approach being better for theme authors/developers, more flexibility ultimately and the ability to use it in other components. I am going to add it as a dependency for these themes!

**Comment by PaulieScanlon** [comment]
@ehowey Nice! 

i'd even say that although it's not at `v1.0.0` yet and i'm probably going to release a few more patch version it is pretty stable. 

The core won't change and i'm not planning on deprecating any of the currently supported providers so all that'll happen is there'll be more components add, and as mentioned before at some point they'll be a more generic component which should making using the plugin a little less fussy! 

I did kinda start a [CHANGELOG](https://github.com/PaulieScanlon/gatsby-mdx-embed/blob/master/%40pauliescanlon/gatsby-mdx-embed/CHANGELOG.md) so i'll do my best to keep that up to date as well just in case you're seeing any problems!

**Comment by osiux** [comment]
Sounds good :) was just asking because of the "wait until is stable", but seems like it's stable enough. Will try to make a PR for it!

@PaulieScanlon One quick, maybe unrelated, question. Haven't played with MDX a lot, but in theory `gatsby-remark-embedder` could also be used on MDX files with `<Flex />` or `<Box />`, no? At the end It would just be something like:

```
<Flex>
    <Box>
        https://twitter.com/PaulieScanlon/status/1222205277065433095
    </Box>
</Flex>
```
In my mind that would compile the link to HTML code and put it inside the `Box`. I'm gonna try it later just to see what happens haha

**Comment by ehowey** [comment]
@osiux Would love to have a PR for this. The main item would be adding it as a dependency in the "core" theme. Let me know if you get stuck, need any help or have questions.

**Comment by PaulieScanlon** [comment]
Hey @osiux Thats a really good point and truth be told i don't know if that would work having not tried it myself.

In theory it should but It might come down to where `gatsby-remark-embedder` is defined as a plugin.

I have a feeling it would need to be defined in `gatsbyRemarkPlugins` which is part of the configuration options in for [gatsby-plugin-mdx](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-mdx#configuration)

Perhaps you could give it a go and let me know?

**Comment by ehowey** [comment]
Implemented! Thanks @osiux live in v0.22.0 of the core theme, just published to npm!
